kmod/patch/livepatch-patch-hook.c

```diff
@@ -209,6 +209,128 @@ static void patch_free_livepatch(struct klp_patch *patch)
 	}
 }
 
+extern struct kpatch_pre_patch_callback __kpatch_callbacks_pre_patch[], __kpatch_callbacks_pre_patch_end[];
+extern struct kpatch_post_patch_callback __kpatch_callbacks_post_patch[], __kpatch_callbacks_post_patch_end[];
+extern struct kpatch_pre_unpatch_callback __kpatch_callbacks_pre_unpatch[], __kpatch_callbacks_pre_unpatch_end[];
+extern struct kpatch_post_unpatch_callback __kpatch_callbacks_post_unpatch[], __kpatch_callbacks_post_unpatch_end[];
+
+static int patch_add_callbacks_to_object(void)
+{
+	struct kpatch_pre_patch_callback *p_pre_patch_callback;
+	struct kpatch_post_patch_callback *p_post_patch_callback;
+	struct kpatch_pre_unpatch_callback *p_pre_unpatch_callback;
+	struct kpatch_post_unpatch_callback *p_post_unpatch_callback;
+	struct patch_object *object;
+
+	for (p_pre_patch_callback = __kpatch_callbacks_pre_patch;
+	     p_pre_patch_callback < __kpatch_callbacks_pre_patch_end;
+	     p_pre_patch_callback++) {
+		object = patch_find_object_by_name(p_pre_patch_callback->objname);
+		if (!object)
+			return -ENOMEM;
+	}
+
+	for (p_post_patch_callback = __kpatch_callbacks_post_patch;
+	     p_post_patch_callback < __kpatch_callbacks_post_patch_end;
+	     p_post_patch_callback++) {
+		object = patch_find_object_by_name(p_post_patch_callback->objname);
+		if (!object)
+			return -ENOMEM;
+	}
+
+	for (p_pre_unpatch_callback = __kpatch_callbacks_pre_unpatch;
+	     p_pre_unpatch_callback < __kpatch_callbacks_pre_unpatch_end;
+	     p_pre_unpatch_callback++) {
+		object = patch_find_object_by_name(p_pre_unpatch_callback->objname);
+		if (!object)
+			return -ENOMEM;
+	}
+
+	for (p_post_unpatch_callback = __kpatch_callbacks_post_unpatch;
+	     p_post_unpatch_callback < __kpatch_callbacks_post_unpatch_end;
+	     p_post_unpatch_callback++) {
+		object = patch_find_object_by_name(p_post_unpatch_callback->objname);
+		if (!object)
+			return -ENOMEM;
+	}
+
+	return 0;
+}
```

Comment:
Is it me or does this function not actually do anything? :-)

Reply:
It's not you... I'm not sure what I was trying to do there, but it doesn't look like anything productive.

Reply:
Ugh, actually there's an important side effect of `patch_find_object_by_name()` and that it calls `patch_alloc_new_object()` which adds to the `patch_objects` list.  If there are no code changes to those same functions, then the callbacks will be dropped as they never make it onto the `patch_objects` list.

Reply:
This is still confusing.  patch_find_object_by_name() should be renamed (patch_find_or_add_object?).  And I suppose patch_add_callbacks_to_object() also needs to be renamed (and maybe commented?) as it doesn't actually add any callbacks.  And why is it split up into two functions anyway (this one and patch_set_klp_callbacks)?